c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -271,7 +271,46 @@ impl<'tcx> TypeChecker<'tcx, '_> {
                     Label::default()
                 })
             }
-
+            Rvalue::Cast(CastKind::PointerFromExposedAddress, ref op, _ty) => {
+                // We support only one case here, which is the case of null pointers
+                // constructed via casts such as `0 as *const T`
+                if let Some(true) = op.constant().cloned().map(util::is_null_const) {
+                    let adt_metadata = &construct_adt_metadata(self.tcx);
+                    self.ltcx.relabel(expect_ty, &mut |lty| {
```

Comment:
I had to think about why relabeling `expect_ty` is a reasonable thing to do here.  I think it works but is a bit of a hack.  Labeling `ty` seems more correct, but the problem is we don't know what `perms` to set for it (we particularly care about `UNIQUE` here).  The correct solution is probably to add an `rvalue_tys` mechanism for borrowck similar to what we have for dataflow, but the approach here should work for now.

Can you add a comment noting that we're only extracting `perms` from `expect_ty`, and otherwise this is conceptually similar to labeling the cast target `ty`?

Reply:
done in ef0a754e